### ISSUE TITLE
scheduler.queue_task() is broken by validation

### ISSUE THREAD

**Issue by kjkuan** [web2py/web2py#1013]
This used to work, but now even the example in the book doesn't work.

```
>>> scheduler.queue_task('demo1', [1,2])
<Row {'errors': {'status': 'Value not allowed', 'start_time': 'Enter date and time as 1963-08-28 14:30:59', 'period': 'Enter an integer greater than or equal to 0', 'sync_output': 'Enter an integer greater than or equal to 0
', 'timeout': 'Enter an integer greater than or equal to 1', 'application_name': 'Enter a value', 'retry_failed': 'Enter an integer greater than or equal to -1', 'repeats': 'Enter an integer greater than or equal to 0'}, 'id
': None, 'uuid': None}>

```

**Comment by kjkuan** [comment]
I suspect it's due to recent changes to the ``validate_and_insert`` method... @gi0baro 
Is this the correct behavior for ``validate_and_insert`` now?

**Comment by niphlod** [comment]
@gi0baro : "my" fault of not adding tests for scheduler (still searching for ways to do multiprocessing tests) but in any case, validate_and_insert  where there's a default on the Field should include that default before passing into the validation stage.

**Comment by gi0baro** [comment]
@niphlod will look on this tomorrow.

**Comment by gi0baro** [comment]
Hmm. I think we should start a discussion about this (here, on pydal repo or on the development group) since the main issue here is 'logical': validation is always done before the *insertion* process, and the default values are applied by the "insert flow". I think the actual behavior of `validate_and_insert` is correct:
- take the values from user
- validate the values with every field's requirements
- if everything was good, call insert
- if some of the values are missing and field have a default, put value in
- finally create record

So, considering https://github.com/web2py/web2py/blob/master/gluon/scheduler.py#L615, under my point of view writing:
```
Field('status', requires=IS_IN_SET(TASK_STATUS), default=QUEUED, writable=False)
```
means you are requiring "on the validation process" that the given value for `'status'` is in the set `TASK_STATUS`. A missing value is `None`, so is not in the set and the validation fails.
Again, the `requires` is intended to be used for validation process, not the insert. In particular, if I want that a value can be missing, and if not has to be into a given set, I would write:
```python
Field('status', requires=IS_EMPTY_OR(IS_IN_SET(TASK_STATUS)), default=QUEUED, writable=False)
```

@niphlod do you see my point?

**Comment by niphlod** [comment]
see it but do not understand the practical case. I don't want in any case that a null value or '' lands in 'status'. But I do want that every value put in there stays in a set. And that if noone sets a value for it, I supply a default one. Doing what you ask (shifting the requires attribute to include IS_EMPTY_OR) doesn't really qualify to make my status field proofed.

**Comment by gi0baro** [comment]
@niphlod actually it does:
- you call validate_and_insert() without status
- pydal runs validation on status with `IS_EMPTY_OR`
- validation is successful because `status` is `None`
- insert collect fields, `status` is missing
- since `status` is missing, pydal write the default value in `status`
- record is created in table

Am I missing something?

**Comment by niphlod** [comment]
but my field can't be empty. No None and no ''  .... frankly it's counter-intuitive to validate with IS_EMPTY_OR something I do want to never be empty in any case, no ?

**Comment by gi0baro** [comment]
@niphlod Yes, but you don't want to be empty in the table, not in validation. Here is the "logical issue":
- you don't want inserts without `status` pass the validation
- you're calling `validate_and_insert` without any `status`

The "real logic" here is that you want:
- inserts that have a value for `status` must be into a specific set
- insert without a value for `status` (which are by definition, *empty*) to be inserted in table with a default value (and this is the *definition* of a default value)
The third clause "the default value has to be in the set" is implicit because you are defining a default value which IS in your set.

In summary, using `IS_EMPTY_OR(IS_IN_SET(yourset))` in combination with a default value means:
- if no value is passed to `validate_and_insert` or the passed value is *empty* (`None`, `''`, ecc..) you will have a new record inserted with the `default` value you provided
- if the given value passed to `validate_and_insert` is in set you will have a new record inserted with the given value
- if the given value passed to `validate_and_insert` is not in the set, validation fails and no record will be created
This means: you will never have a null value into the column specified.

**Comment by niphlod** [comment]
uhm. even if it has its own logic, it's ugly and counter-intuitive. All that it takes for validate_and_insert (or validate_and_update) is to update the object before passing to the validation, simulating how the record will be created in the database, respectively with "default" and "update" values.
Then you can have a 'status' with requires=IS_IN_SET() without a default if you want the value to be explicitely passed.

**Comment by gi0baro** [comment]
I think is counter-intuitive because we're mixing *database level* requirements with *validation level* requirements. What you *require* in validation is not the same "layer" of what you require on the database (and I think this is not IMHO).

Considering the *validation* process, as an application developer, I don't see any reason for the framework to *alter* the input before the validation.
You're practically saying that a `default` or an `update` value will somehow *mysteriously* edit what I've written in the `requires` parameter :-\

**Comment by kjkuan** [comment]
What's wrong with validating and using the default values if the values for some fields are not provided? Isn't that what default values are for? Something like:
```
diff --git a/pydal/objects.py b/pydal/objects.py
index 3415ca8..cc62b3a 100644
--- a/pydal/objects.py
+++ b/pydal/objects.py
@@ -714,7 +714,13 @@ class Table(Serializable, BasicStorage):
         response.id, response.errors = None, Row()
         new_fields = copy.copy(fields)
         for fieldname in self.fields:
-            value, error = self[fieldname].validate(fields.get(fieldname))
+            if fieldname in fields:
+                fieldvalue = fields[fieldname]
+            else:
+                fieldvalue = self[fieldname].default
+                if callable(fieldvalue):
+                    fieldvalue = fieldvalue()
+            value, error = self[fieldname].validate(fieldvalue)
             if error:
                 response.errors[fieldname] = "%s" % error
             else:
```

Besides, I wonder what the impact might be if we change the old behavior of ``validate_and_insert``, no one complained about the old behavior, or is there?

**Comment by niphlod** [comment]
not really. IMHO default and update should map to the creation of the field in the backend with a costraint (and optionally prepopulate forms when shown the first time, but that's another matter). 
Given we don't have that, imho it's a lot more intuitive having

Field('status', requires=IS_IN_SET(...), default='xyz')

rather than 

Field('status', requires=IS_EMPTY_OR(IS_IN_SET(...), default='xyz')

for something I want to never be empty or None.

**Comment by gi0baro** [comment]
@niphlod @kjkuan AFAIK no one web framework override input values with defaults/updates on validation. You usually use callbacks for this.
IMHO is not a logical behavior. If we use this behavior will means:
- `Field('status', requires=IS_IN_SET(...), default='xyz')` and `Field('status', requires=IS_EMPTY_OR(IS_IN_SET(...), default='xyz')` are the same
- `Field('foo', requires=IS_NOT_EMPTY(), default='bar')` will pass the validation even if no value is given by user
- `Field('asd', requires=IS_IN_SET(['lol', 'rotfl']), default='rotfl')` wont validate the condition "value is in the set I've declared" when the value is empty
- To ensure a value is in a set and is not empty I should write `Field('asd', requires=[IS_NOT_EMPTY(), IS_IN_SET(['lol', 'rotfl'])], default='rotfl')`
- Writing `IS_IN_SET` with a `default` value won't really validate that "value given by a user is in the set" because `None` will be overwritten but wasn't in the set.

Personally, I think this is too much more counter-intuitive of thinking "the value given should be in the set and when it's empty will be overwritten" -> `IS_EMPTY_OR(IS_IN_SET())`..

**Comment by gi0baro** [comment]
Addition:
"the value given should be in the set and when it's empty will be overwritten" assumes you are allowing empty values. If you don't allow empty values you also don't have to specify a `default` because user must choose between the ones in the set.

**Comment by kjkuan** [comment]
I'd agree with @gi0baro if we are only doing validation. But the method is called, ``validate_and_insert``, and from the documentation of this method in the book, the emphasis to me seems to be more on the *insert* side. That is, it's just like ``insert``, only before really inserting, it validates the values that are going to be inserted and bails out if they don't validate.

I wonder what's @mdipierro 's opinion on this.

**Comment by gi0baro** [comment]
@kjkuan You're right: I'm assuming `validate_and_insert` should behave like writing

```python
rv = db.table.validate(input)
if not rv.errors:
    db.table.insert(input)
```

I would also hear @mdipierro on this.

**Comment by niphlod** [comment]
funny. insert(input) won't fail at all (it skips validation) but still the 'QUEUED' would be in the 'status' column.

**Comment by gi0baro** [comment]
@niphlod as intended. no validation is performed by `insert` and defaults are injected by `_insert`..

**Comment by niphlod** [comment]
yep. but AFTER insertion the record is perfectly valid even if the validate_and_insert how it's implemented right now fails.

**Comment by gi0baro** [comment]
@niphlod yep. Because in this specific case is the same as not validating the input.

**Comment by niphlod** [comment]
dunno. IMHO it's counterintuitive that the result of an insertion is a record that complies fully with the validation but a pre-validation step fails.

**Comment by mdipierro** [comment]
@gi0baro complains that: "Field('foo', requires=IS_NOT_EMPTY(), default='bar') will pass the validation even if no value is given by user". This actually a feature. :-)
If a value is passed, even if None, is should not be ignored and go through the validation process. If no value is actually specified, the default value should be used. This is a requirement and we do not want to break book examples. So I am with @niphlod on this one.

**Comment by mdipierro** [comment]
I believe this is fixed. Any objection to close it?
